Add manufacturer information

kornefalk opened the pull request:
Adding manufacturer information to the Maven Plugin.

This is very useful for companies and organizations that can add information to the SBOM files about the source and how to contact the application vendor. This information is optional and one configuration example is included. If this information is not included, the plugin works as before without the manufacturer section. 

When using a centralized parent POM file, all artefacts can create a manufacturer information. This kind of information is not often changed withing the company or organization.

_Some checks of the manufacturer  information is available, should maybe be moved to the core project._

ppkarwasz commented:
I would rather reuse what is already there in the POM file, e.g.:

```xml
<organization>
  <name>The Apache Software Foundation</name>
  <url>https://apache.org</url>
</organization>
<developers>
  <developer>
    <name>Apache Logging Services</name>
    <email>security@logging.apache.org</email>
    <roles>
      <role>security_contact</role>
    </roles>
    <properties>
      <postalAddress>
        The Apache Software Foundation
        1000 N West Street, Suite 1200
        Wilmington, DE 19801
        U.S.A.
      </postalAddress>
    </properties>
  </developer>
</developers>
```

kornefalk commented:
The `organization` section is also relevant to the manufacturer information. All fields in the *developers` properties section is _"This element is where any other properties about the person goes."_ 

I can rewrite this PR with usage of the `organization` and `developers`, and it will not by default add fields that does not match one to one like postal address.

ppkarwasz commented on src/test/java/org/cyclonedx/maven/BaseCycloneDxMojoTest.java:
```diff
@@ -0,0 +1,132 @@
+    @DisplayName("Verify that the default configuration does not have manufacturer information ")
+    void createManufacturer() {
+        BaseCycloneDxMojoImpl mojo = new BaseCycloneDxMojoImpl();
+        OrganizationalEntity manufacturer = mojo.createManufacturer(null, null);
+        assertNotNull(manufacturer);
```
Should this be `null` if there is no data?

kornefalk replied:
Security check if you have a very minimal POM file without anything. Could be removed, since checks are in place before calling the tested function, but I didn't see any unit testing at all to the `BaseCycloneDxMojo`, only some testing with POM files running with the super classes. Writing unit test for `execute()` function will take lot of time.

```java
                if (organization != null || (developers != null && !developers.isEmpty())) {
                    metadata.setManufacturer(createManufacturer(organization, developers));
                }
```

ppkarwasz commented on src/test/java/org/cyclonedx/maven/BaseCycloneDxMojoTest.java:
```diff
@@ -0,0 +1,132 @@
+     * @throws NoSuchFieldException If the field does not exist
+     * @throws IllegalAccessException If the value is not able to be modified
+     */
+    public static void setParentParameter(Object cc, String fieldName, Object value)
+        throws NoSuchFieldException, IllegalAccessException {
+        Field field = cc.getClass().getSuperclass().getDeclaredField(fieldName);
+        field.setAccessible(true);
+        field.set(cc, value);
+    }
```
Is this used anywhere?

kornefalk replied:
Sorry forgot to remove from the first PR, not used any more. It was used to set the private variable `manufacturer` before.

kornefalk commented:
With the given example given by @ppkarwasz , the output will be

```json
{
 "metadata" : {
    "authors" : [
      {
        "name" : "Apache Logging Services Security Team",
        "email" : "security@logging.apache.org"
      }
    ],
    "manufacturer" : {
      "name" : "The Apache Software Foundation",
      "url" : [
        "https://apache.org",
        "https://logging.apache.org/security.html"
      ],
      "contact" : [
        {
          "name" : "Apache Logging Services Security Team",
          "email" : "security@logging.apache.org"
        }
      ]
    }
}
```
Parsing a property field with postal address into ZIP code, street address, city, country etc is very complicated to make it work with over 100 countries who has different ways of writing the postal information. The `manufacturer` section  several fields witch is very hard to get it correct.

The section `manufacturer.contact` and `authors` in the metadata, has identical formats, which is little confusing.

hboutemy commented:
from the code update, I see now that we have both manufacturer https://cyclonedx.org/docs/1.6/json/#metadata_manufacturer
and authors https://cyclonedx.org/docs/1.6/json/#metadata_authors
with data extracted from POM organization https://maven.apache.org/ref/3.9.9/maven-model/maven.html#class_organization and https://maven.apache.org/ref/3.9.9/maven-model/maven.html#class_developer

but no documentation any more: I'd like to have a page on this, like https://cyclonedx.github.io/cyclonedx-maven-plugin/external-references.html
this will help users getting suprising values and discovering what they have in their effective POM (with everything they inherited, eventually not really being aware of it: I've seen that a lot with this plugin, where people complain about the plugin instead of trying to fix their data)

hboutemy commented:
reading CycloneDX doc for manufacturer and authors
> The organization that created the BOM.
> Manufacturer is common in BOMs created through automated processes. BOMs created through manual means may have `@.authors` instead.

> The person(s) who created the BOM.
> Authors are common in BOMs created through manual processes. BOMs created through automated means may have @.manufacturer instead.

and thinking at real world content for developers in POMs, where there is a description of many individuals, and not always accurate

perhaps we should avoid the authors part: I fear this will just add noise in real world

ppkarwasz commented:
> this will help users getting suprising values and discovering what they have in their effective POM (with everything they inherited, eventually not really being aware of it: I've seen that a lot with this plugin, where people complain about the plugin instead of trying to fix their data)

This is how we discovered that all the effective `<url>` metadata in our POMs was bogus, because of inheritance.

> and thinking at real world content for developers in POMs, where there is a description of many individuals, and not always accurate
> 
> perhaps we should avoid the authors part: I fear this will just add noise in real world

Maybe this will give developers an opportunity to clean up their POMs. For projects that are not maintained by a single person, it doesn't make sense to add anything else than a link to the maintainer team in `<developers>` (and it doesn't make sense to use `@author` Javadoc tags).

Projects with hundreds of contributors and dozens of past maintainers should probably add a single **long term** contact information to the `<developers>` section. However, to limit the impact of this PR and prevent users from sending security reports to public mailing lists and similar, I would include in `$.authors` only the developers with a specific `<role>`, e.g. "Security Team".

kornefalk commented:
- Move the manufacturer section from metadata to metadata/component/manufacturer due to deprecation
- Modified the creation of manufacturer data to work with aggregation.
- Re-fractured  the code
- Removed call to  `setAuthors` since it is not needed

Example output:
```json
{
  "metadata" : {
    "component" : {
      "type" : "library",
...
      "manufacturer" : {
        "name" : "The Apache Software Foundation",
        "url" : [
          "https://apache.org",
          "https://logging.apache.org/security.html"
        ],
        "contact" : [
          {
            "name" : "Apache Logging Services Security Team",
            "email" : "security@logging.apache.org"
          }
        ]
      },
...
    }
   }
}
```

kornefalk commented:
One option with the `<role>`, could be to have one configuration parameter, that determines which developer should be includes. Like:
```xml
   <configuration>
      <includeRoles>Security Team</includeRoles>
   <configuration>
```
And implement it as a `List<String>`, so you can specify one or more roles to include. It could be defaulted to `Security Team`.
If so, this new configuration parameter needs to be documented. What do you say @hboutemy ?

ppkarwasz commented:
Yes, it seems that `$.metadata.manufacturer` is not the right element to fill with data from the POM file. What about `$.components[*].manufacturer` (https://cyclonedx.org/docs/1.6/json/#components_items_manufacturer)?
